File: src/main/java/br/com/kronos/backend/aplicacao/calendario/Calendario.java

package br.com.kronos.backend.aplicacao.calendario;

import static br.com.kronos.backend.aplicacao.validador.Validacoes.exigirNaoNulo;
import static br.com.kronos.backend.aplicacao.validador.Validacoes.exigirTamanho;

import java.time.LocalDate;

import lombok.Builder;
import lombok.Getter;

@Getter
public class Calendario {
    
    private long id;
    private Integer numero;
    private Integer ano;
    private LocalDate dataInicio;
    private LocalDate dataFinal;
    private LocalDate dataInicioAtribuicaoCredito;
    private LocalDate dataFinalAtribuicaoCredito;
    private LocalDate dataInicioAnoLetivo;
    private LocalDate dataFinalAnoLetivo;
    private String descricao;
    private LocalDate dataAprovacao;
    private Long idInstituicao;
    private Long idFuncionarioAprovacao;
    private Integer idTipoSituacaoCalendario;
    private LocalDate dataConclusao;
    private Long idFuncionarioElaboracao;
  
	@Builder
    public Calendario(long id, Integer numero, Integer ano, LocalDate dataInicio, LocalDate dataFinal, LocalDate dataInicioAtribuicaoCredito, 
                      LocalDate dataFinalAtribuicaoCredito, LocalDate dataInicioAnoLetivo, LocalDate dataFinalAnoLetivo, 
                      String descricao, LocalDate dataAprovacao, Long idInstituicao, Long idFuncionarioAprovacao, 
                      Integer idTipoSituacaoCalendario,  LocalDate dataConclusao, Long idFuncionarioElaboracao) {

        this.id = id;
        this.numero = numero; 
        this.ano = exigirNaoNulo(ano, "Ano"); 
        this.dataInicio = exigirNaoNulo(dataInicio, "Data de início"); 
        this.dataFinal = exigirNaoNulo(dataFinal, "Data final"); 
        this.dataInicioAtribuicaoCredito = exigirNaoNulo(dataInicioAtribuicaoCredito, "Data de início de matricula"); 
        this.dataFinalAtribuicaoCredito = exigirNaoNulo(dataFinalAtribuicaoCredito, "Data final de matricula"); 
        this.dataInicioAnoLetivo = exigirNaoNulo(dataInicioAnoLetivo, "Data de inicio do ano letivo"); 
        this.dataFinalAnoLetivo = exigirNaoNulo(dataFinalAnoLetivo, "Data final do ano letivo"); 
        this.descricao = exigirTamanho(descricao, "Descrição", 1, 500); 
        this.dataAprovacao = dataAprovacao;
        this.idInstituicao = exigirNaoNulo(idInstituicao, "Instituição"); 
        this.idFuncionarioAprovacao = idFuncionarioAprovacao; 
        this.idTipoSituacaoCalendario = id == 0 ? EnumTipoSituacaoCalendario.EM_ELABORACAO.id() : 
        		exigirNaoNulo(idTipoSituacaoCalendario, "Situação do calendário"); 
        this.dataConclusao = dataConclusao;
        this.idFuncionarioElaboracao = idFuncionarioElaboracao;
	}
}

File: poc-kronos-frontend/src/model/calendario/Calendario.js
class Calendario {
  static EM_ELABORACAO() {
    return 1;
  }
  static CONCLUIDO() {
    return 2;
  }
  static APROVADO() {
    return 3;
  }
  constructor(dados) {
    this.id = dados ? dados.id : null;
    this.numero = dados ? dados.numero : null;
    this.ano = dados ? dados.ano : null;
    this.dataInicio = dados ? dados.dataInicio : null;
    this.dataFinal = dados ? dados.dataFinal : null;
    this.dataInicioAtribuicaoCredito = dados
      ? dados.dataInicioAtribuicaoCredito
      : null;
    this.dataFinalAtribuicaoCredito = dados
      ? dados.dataFinalAtribuicaoCredito
      : null;
    this.dataInicioAnoLetivo = dados ? dados.dataInicioAnoLetivo : null;
    this.dataFinalAnoLetivo = dados ? dados.dataFinalAnoLetivo : null;
    this.descricao = dados ? dados.descricao : null;
    this.dataAprovacao = dados ? dados.dataAprovacao : null;
    this.idInstituicao = dados ? dados.idInstituicao : null;
    this.idFuncionarioAprovacao = dados ? dados.idFuncionarioAprovacao : null;
    this.nomeFuncionarioAprovacao = dados
      ? dados.nomeFuncionarioAprovacao
      : null;
    this.idTipoSituacaoCalendario = dados
      ? dados.idTipoSituacaoCalendario
      : null;
    this.nomeTipoSituacaoCalendario = dados
      ? dados.nomeTipoSituacaoCalendario
      : null;
    this.dataConclusao = dados ? dados.dataConclusao : null;
    this.idFuncionarioElaboracao = dados ? dados.idFuncionarioElaboracao : null;
    this.nomeFuncionarioElaboracao = dados
      ? dados.nomeFuncionarioElaboracao
      : null;
    this.idsCursos = dados ? dados.idsCursos : null;
  }

  static validations(required, numeric, maxLength, dataFimMenor) {
    return {
      ano: { required, numeric },
      dataInicio: { required },
      dataFinal: { required, dataFimMenor: dataFimMenor("dataInicio") },
      dataInicioAtribuicaoCredito: { required },
      dataFinalAtribuicaoCredito: {
        required,
        dataFimMenor: dataFimMenor("dataInicioAtribuicaoCredito")
      },
      dataInicioAnoLetivo: { required },
      dataFinalAnoLetivo: {
        required,
        dataFimMenor: dataFimMenor("dataInicioAnoLetivo")
      },
      descricao: { required, maxLength: maxLength(500) },
      idsCursos: { required }
    };
  }

  static validationMessages() {
    return {
      ano: {
        required: "Ano é obrigatório",
        numeric: "Somente números"
      },
      dataInicio: {
        required: "Data Início é obrigatório"
      },
      dataFinal: {
        required: "Data Final é obrigatória",
        dataFimMenor: "Data Final não pode ser menor que a Data Inicial"
      },
      dataInicioAtribuicaoCredito: {
        required: "Data Início Atribuição Crédito é obrigatório"
      },
      dataFinalAtribuicaoCredito: {
        required: "Data Final Atribuição Crédito é obrigatória",
        dataFimMenor:
          "Data Final Atribuição Crédito não pode ser menor que a Data Início Atribuição Crédito"
      },
      dataInicioAnoLetivo: {
        required: "Data Início Ano Letivo é obrigatório"
      },
      dataFinalAnoLetivo: {
        required: "Data Final Ano Letivo é obrigatória",
        dataFimMenor:
          "Data Final Ano Letivo não pode ser menor que a Data Início Ano Letivo"
      },
      descricao: {
        required: "Descrição é obrigatória"
      },
      idsCursos: {
        required: "Curso é obrigatório"
      }
    };
  }
}
export default Calendario;
